invenio_records_resources/services/records/queryparser/suggest.py

```diff
@@ -51,7 +51,7 @@ class MySearchOptions(SearchOptions):
     def __init__(self, identity=None, extra_params=None, **kwargs):
         """Constructor."""
         super().__init__(identity=identity, extra_params=extra_params)
-        self.extra_params.setdefault("type", "bool_prefix")
+        self.extra_params.setdefault("type", "best_fields")
```

Comment:
So I was curious to understand better why this would give better results :thinking: ... I mean if you've observed it, I trust it. It's wasn't very clear to me *why* though.

From reading https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-multi-match-query.html it seems like `bool_prefix` does combine scores too (it sums the score across each field it targets). `best_fields` picks the highest score among the fields it targets (by assigning a higher score to a field, the more it contains the search terms). 

However, if I understand correctly it's the `prefix` part/query (well, its absence now) that makes a difference here; now, finding the term prefixing something in the field doesn't score higher. The `match` query used by `best_fields` under the hood (as opposed to multi `term` + `prefix`) probably gives a higher score to fields that have higher ratios of total characters to matched characters. So something like:

```
Search: "fr"
Over: "French", "Cajun French", "Old French (842-ca. 1400)", "Eastern Frisian"
```

would presumably score "French" higher now because it matches a higher ratio of the content...

Is that right?

Reply:
Yes, you're right. I was over-simplifying. The `bool_prefix` search does look at summed scores across the fields. In our case because the fields being searched are manipulations of the one short vocabulary term string (e.g.,  "subject" and the derived "subject._2gram" and "subject._3gram"), I don't think the scores are going to vary between the fields. 

There is also some scoring involved in the search in individual fields, but it doesn't seem to be very meaningful for us. A "bool_prefix" search converts the last term in the search string to a prefix search. Prefix searches seem not to be weighted. They score a 1 if any word in the field begins with the prefix string, regardless of where that word comes in the field or (as you say) the proportion of the field it occupies. So in a vocabulary search, any search string that's a single term will produce a result set with identical scores. This is what I was finding in my initial tests with "bool_prefix": all of the results had a score of 100.

As you say, when the search string extends to more than one term, then only the last term is converted to a prefix search. The others are converted to exact term searches. Here there *is* a slight variation in scoring. But the only things that I've so far seen to affect the score are

1) whether the complete field text is a match to the exact term, and if not then
2) whether a longer field contains at least one exact match in a longer text, and
3) the number of exact matches in that longer text 

As you point out, where the whole field isn't an exact match, there don't appear to be any variation for, e.g., position of the match in the text or proportion of the text occupied by the match. There are likely factors I'm missing here, but the result is only a limited set of different scores that aren't that helpful. (The docs recommend against term searches for text fields.)  

For multi-term search strings, the bool_prefix search constructs a boolean search combining these prefix and term searches with the "should" operator. So a search for "computer programming" is converted to:
```
{
  "query": {
    "bool" : {
      "should": [
        { "term": { "subject": "computer" }},
        { "prefix": { "subject": "programming"}}
      ]
    }
  }
}
```
I haven't dug into how the "should" operator works here, and in my tests there seem to be some complexities that produce score variations. But they don't tend to be useful to us, and are in some cases counter-intuitive. 

The upshot is what you've said: that the factors we intuitively expect to be used--things like proportion of the text matched and position of the match in the text--aren't taken into account by "bool_prefix". And most problematic, there's *no* variation in scoring at all for single term queries.

And (again, as you point out) the "match" queries fit those intuitive expectations much better.

Reply:
I may have responded at the wrong level of the thread! Hopefully you can see that I responded, though. Thanks for approving the change.

Reply:
Thanks so much for that. I think that makes sense to me. Elastic/Open search is a beast. Someone else at CERN (or other collaborator) should give it a last check before we merge it, but that all sounds good to me. I am glad we have some general sense of what is going.